tap_agent/src/tap/account.rs

```diff
@@ -0,0 +1,846 @@
+            allocation_id,
+            sender,
+            required_checks.clone(),
+        );
+        let rav_storage_adapter = RAVStorageAdapter::new(pgpool.clone(), allocation_id, sender);
+        let tap_manager = TapManager::new(
+            tap_eip712_domain_separator.clone(),
+            escrow_adapter,
+            receipt_checks_adapter,
+            rav_storage_adapter,
+            receipt_storage_adapter,
+            required_checks,
+            0,
+        );
+        Self {
+            inner: Arc::new(Inner {
+                pgpool,
+                tap_manager,
+                allocation_id,
+                sender,
+                sender_aggregator_endpoint,
+                unaggregated_fees: Arc::new(Mutex::new(UnaggregatedFees::default())),
+                state: Arc::new(Mutex::new(State::Running)),
+                config,
+            }),
+            rav_requester_task: Arc::new(Mutex::new(None)),
+        }
+    }
+
+    pub async fn handle_new_receipt_notification(
+        &self,
+        new_receipt_notification: NewReceiptNotification,
+    ) {
+        // If we're in the last rav pending state, we don't want to process any new receipts.
+        if self.state().await == State::LastRavPending {
+            error!(
+                "Received a new receipt notification for allocation {} and sender {} while \
+                the last RAV is pending. This should not have happened since this allocation \
+                and/or sender is not eligible anymore.",
+                self.inner.allocation_id, self.inner.sender
+            );
+            return;
+        }
+
+        let mut unaggregated_fees = self.inner.unaggregated_fees.lock().await;
+
+        // Else we already processed that receipt, most likely from pulling the receipts
+        // from the database.
+        if new_receipt_notification.id > unaggregated_fees.last_id {
+            unaggregated_fees.value = unaggregated_fees
+                .value
+                .checked_add(new_receipt_notification.value)
+                .unwrap_or_else(|| {
+                    // This should never happen, but if it does, we want to know about it.
+                    error!(
+                        "Overflow when adding receipt value {} to total unaggregated fees {} for \
+                        allocation {} and sender {}. Setting total unaggregated fees to u128::MAX.",
+                        new_receipt_notification.value,
+                        unaggregated_fees.value,
+                        new_receipt_notification.allocation_id,
+                        new_receipt_notification.sender_address
+                    );
+                    u128::MAX
+                });
+            unaggregated_fees.last_id = new_receipt_notification.id;
+
+            let mut rav_requester_task = self.rav_requester_task.lock().await;
+            // TODO: consider making the trigger per sender, instead of per (sender, allocation).
```

Comment:
Yes. Makes the code slightly more complex because the RAVs still have to be per (sender, allocation).

Would it make sense to have a strategy for which allocation should be aggregated first, instead of doing all of them?  For example first collect the allocation that has the most unaggregated fees?

Reply:
Moved to #93